### ISSUE TITLE
word2vec install on Windows 8.1 (and Windows XP) 32-bit

### ISSUE THREAD

**Issue by dineshbvadhia** [danielfrg/word2vec#12]
Install on Windows XP and Windows 8.1 caused the same error. Tried both:

> pip install -U word2vec, and
> "python setup.py install" using the zip'ed download

Anaconda 2.1 32-bit.

```
C:\Documents and Settings\dinesh>pip install -U word2vec
Collecting word2vec
  Downloading word2vec-0.6.7.tar.gz
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "c:\docume~1\dinesh\locals~1\temp\pip-build-d9dqjn\word2vec\setup.py"
, line 17, in <module>
        subprocess.call(['make', '-C', 'word2vec-c'])
      File "C:\Anaconda\lib\subprocess.py", line 522, in call
        return Popen(*popenargs, **kwargs).wait()
      File "C:\Anaconda\lib\subprocess.py", line 710, in __init__
        errread, errwrite)
      File "C:\Anaconda\lib\subprocess.py", line 958, in _execute_child
        startupinfo)
    WindowsError: [Error 2] The system cannot find the file specified
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):

      File "<string>", line 20, in <module>

      File "c:\docume~1\dinesh\locals~1\temp\pip-build-d9dqjn\word2vec\setup.py"
, line 17, in <module>

        subprocess.call(['make', '-C', 'word2vec-c'])

      File "C:\Anaconda\lib\subprocess.py", line 522, in call

        return Popen(*popenargs, **kwargs).wait()

      File "C:\Anaconda\lib\subprocess.py", line 710, in __init__

        errread, errwrite)

      File "C:\Anaconda\lib\subprocess.py", line 958, in _execute_child

        startupinfo)

    WindowsError: [Error 2] The system cannot find the file specified

    ----------------------------------------
←[31m    Command "python setup.py egg_info" failed with error code 1 in c:\docum
e~1\dinesh\locals~1\temp\pip-build-d9dqjn\word2vec←[0m
..............................................................................................................................................
C:\Documents and Settings\dinesh\My Documents\Downloads\word2vec-master>python s
etup.py install
Traceback (most recent call last):
  File "setup.py", line 17, in <module>
    return_code = subprocess.call(['make', '-C', 'word2vec-c'])
  File "C:\Anaconda\lib\subprocess.py", line 522, in call
    return Popen(*popenargs, **kwargs).wait()
  File "C:\Anaconda\lib\subprocess.py", line 710, in __init__
    errread, errwrite)
  File "C:\Anaconda\lib\subprocess.py", line 958, in _execute_child
    startupinfo)
WindowsError: [Error 2] The system cannot find the file specified
```

**Comment by danielfrg** [comment]
Yes, the pkg does not work on windows because i rely on make which is not available there.

**Comment by dineshbvadhia** [comment]
Ah! Would the combination of MinGW (which includes a pthread library) +MySYS (which includes 'make') be sufficient for Windows support?

**Comment by danielfrg** [comment]
Haven't used windows in a really long time but i would love to hear if you try :)

**Comment by dineshbvadhia** [comment]
Tried 3 approaches with identical (failed) results. See below.

1. Install MinGW and MSYS under Anaconda root. 
1a. MinGW is automatically installed with Anaconda but NOT MSYS. 
1b. Install MSYS under MinGW (http://sourceforge.net/projects/mingw/files/)
1c. Add to PATH statement C:\Anaconda\MinGW\bin; C:\Anaconda\MinGW\mysys\1.0\bin; 

2. Install MinGW and MSYS under system root (ie. C:\) (http://sourceforge.net/projects/mingw/files/). 
2a. Install MSYS under MinGW.
2b. Add to PATH statement C:\MinGW\bin; C:\MinGW\mysys\1.0\bin; 

3. Install 3rd party GCC tools bundle TDM-GCC-32 (http://tdm-gcc.tdragon.net/)
3a. Add to PATH statement C:\TDM-GCC-32\bin;
3b. Find "make" in C:\TDM-GCC-32\bin as mingw32-make.exe. Either make copy and rename to "make.exe" or add to word2vec.c code to differentiate between Windows and linux.

4. https://msysgit.github.io/ is another option.
4a. Untested


Open terminal and navigate to word2vec folder. Type

> "python setup.py install" 

First 3 approaches above produced identical output:


C:\Downloads\word2vec-master>python setup.py install
make: Entering directory 'C:/Downloads/word2vec-master/word2vec-c'
gcc word2vec.c -o ../bin/word2vec -lm -pthread -O3 -Wall -march=native -funroll-loops  -Wno-unused-result
word2vec.c:46:1: error: unknown type name 'clock_t'
 clock_t start;
 ^
word2vec.c: In function 'LearnVocabFromTrainFile':
word2vec.c:279:7: warning: unknown conversion type character 'l' in format [-Wformat=]
       printf("%lldK%c", train_words / 1000, 13);
       ^
word2vec.c:279:7: warning: format '%c' expects argument of type 'int', but argument 2 has type 'long long int' [-Wformat=]
word2vec.c:279:7: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c:291:5: warning: unknown conversion type character 'l' in format [-Wformat=]
     printf("Vocab size: %lld\n", vocab_size);
     ^
word2vec.c:291:5: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c:292:5: warning: unknown conversion type character 'l' in format [-Wformat=]
     printf("Words in train file: %lld\n", train_words);
     ^
word2vec.c:292:5: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c: In function 'SaveVocab':
word2vec.c:301:3: warning: unknown conversion type character 'l' in format [-Wformat=]
   for (i = 0; i < vocab_size; i++) fprintf(fo, "%s %lld\n", vocab[i].word, voca
b[i].cn);
   ^
word2vec.c:301:3: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c: In function 'ReadVocab':
word2vec.c:320:5: warning: unknown conversion type character 'l' in format [-Wformat=]
     fscanf(fin, "%lld%c", &vocab[a].cn, &c);
     ^
word2vec.c:320:5: warning: format '%c' expects argument of type 'char *', but argument 3 has type 'long long int *' [-Wformat=]
word2vec.c:320:5: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c:325:5: warning: unknown conversion type character 'l' in format [-Wformat=]
     printf("Vocab size: %lld\n", vocab_size);
     ^
word2vec.c:325:5: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c:326:5: warning: unknown conversion type character 'l' in format [-Wformat=]
     printf("Words in train file: %lld\n", train_words);
     ^
word2vec.c:326:5: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c: In function 'InitNet':
word2vec.c:341:3: warning: implicit declaration of function 'posix_memalign' [-Wimplicit-function-declaration]
   a = posix_memalign((void **)&syn0, 128, (long long)vocab_size * layer1_size *
 sizeof(real));
   ^
word2vec.c: In function 'TrainModelThread':
word2vec.c:366:36: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
   unsigned long long next_random = (long long)id;
                                    ^
word2vec.c:368:3: error: unknown type name 'clock_t'
   clock_t now;
   ^
word2vec.c:372:50: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
   fseek(fi, file_size / (long long)num_threads * (long long)id, SEEK_SET);
                                                  ^
word2vec.c:378:9: warning: implicit declaration of function 'clock' [-Wimplicit-function-declaration]
         now=clock();
         ^
word2vec.c:381:63: error: 'CLOCKS_PER_SEC' undeclared (first use in this function)
          word_count_actual / ((real)(now - start + 1) / (real)CLOCKS_PER_SEC *1000));
                                                               ^
word2vec.c:381:63: note: each undeclared identifier is reported only once for each function it appears in
word2vec.c:413:54: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
       fseek(fi, file_size / (long long)num_threads * (long long)id, SEEK_SET);
                                                      ^
word2vec.c: In function 'TrainModel':
word2vec.c:561:5: warning: unknown conversion type character 'l' in format [-Wformat=]
     fprintf(fo, "%lld %lld\n", vocab_size, layer1_size);
     ^
word2vec.c:561:5: warning: unknown conversion type character 'l' in format [-Wformat=]
word2vec.c:561:5: warning: too many arguments for format [-Wformat-extra-args]
word2vec.c: In function 'TrainModelThread':
word2vec.c:542:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
makefile:11: recipe for target 'word2vec' failed
make: *** [word2vec] Error 1
make: Leaving directory 'C:/Downloads/word2vec-master/word2vec-c'


Note that clock_t start (ln46 in word2vec.c) isn't defined. 

Btw, I found TDM-GCC-32 the easiest to setup and use.

**Comment by dineshbvadhia** [comment]
Update on previous note. For Windows 64-bit, MinGW (or TDM-GCC) is not the best route.

Using the new'ish "Microsoft Visual C++ Compiler for Python 2.7" is the way to go for Python 2.7.

Install Anaconda-2.1.0-Windows-x86_64
Install pywin32-219.win-amd64-py2.7 http://sourceforge.net/projects/pywin32/files/pywin32/Build%20219/

conda update setuptools (>=6.0) and pip (>=6.0)

Install "Microsoft Visual C++ Compiler for Python 2.7" http://www.microsoft.com/en-gb/download/details.aspx?id=44266
Remove all instances of MinGW (or TDM-GCC) outside of Anaconda.
Remove all directions to MinGW (or TDM-GCC) from the PATH statement.

That's all for now.

**Comment by danielfrg** [comment]
Thanks for the updates. There is another solution that might solve the problem on windows (and any other platform), building a conda pkg for word2vec. I thought the idea a long time ago but never did it since it will constrain people to use anaconda. In any case is something i need to solve, eventually, one way or another.

**Comment by danielfrg** [comment]
@dineshbvadhia I just uploaded a change to the `setup.py` so `make` is no longer needed, just gcc should be needed.

I am pretty sure it still doesn't work on windows since the compiler flags are probably will be very different and windows libraries will need to be included but with this change it will be possible to add an if in the `setup.py` and add the correct command.

**Comment by danielfrg** [comment]
I just released a very experimental windows support based on this port: https://github.com/zhangyafeikimi/word2vec-win32

I have also uploaded some wheel packages that were compiled on appveyor. I tested that the commands work but the trained models were not as expected.

I am not a windows user so this is probably as close as i can help people but hopefully this get people closer to using the library there.

I am closing as a partial nofix but new issues and more ideas on how to improve this are welcome.
